### ISSUE TITLE
runBenchmark hangs on macOS

### ISSUE THREAD

**Issue by doomrobo** [bos/criterion#195]
There's some weird platform-specific behavior that I ran into recently. Using the standard `defaultMain` works just fine on macOS, but if I call `runBenchmark` (exposed by `Criterion.Measurement`) directly, my benchmarks don't terminate after >30min of running. Even weirder, when I call `defaultMain` *followed by* a `runBenchmark`, they both terminate.

I wasn't able to replicate this on my Linux machine, but I was able to replicate it on 2 macOS machines (both High Sierra). I also observed it for criterion versions 1.4.0 and 1.1.4.

I have a minimal testcase in a repo [here](https://github.com/doomrobo/criterion-hang-test). This is all the code necessary. It just runs a benchmark on the `id` function.

```haskell
module Main where

import qualified Criterion as C
import qualified Criterion.Main as C
import qualified Criterion.Measurement as C

main :: IO ()
main = do
  let b =  C.nf id (0 :: Int)
  -- If you uncomment the line below, the program will terminate within 30s
  --C.defaultMain [C.bench "id" b]
  x <- C.runBenchmark b 0.0
  print x
```

Any corrections or workarounds would be much appreciated.

**Comment by RyanGlScott** [comment]
Quite strange. I don't have access to macOS, so there's not much I can do on my end to debug this, unfortunately. All I can offer is advice on where to look:

* Does the optimization setting affect this? i.e., does it make a difference whether the program is compiled with `-O0` or `-O`?
* Have you tried putting trace statements inside of `criterion` to narrow down which line of code is hanging?

**Comment by doomrobo** [comment]
I figured it out. The loop inside `runBenchmark` calls `getTime` repeatedly to gauge how long the bench is taking. But `runBenchmark` never calls `initializeTime`! I think this is where the platform-specific behavior came from. Compare the C code for timing functions:

Linux version:
```C
void criterion_inittime(void)
{
}

double criterion_gettime(void)
{
    struct timespec ts;

    clock_gettime(CLOCK_MONOTONIC, &ts);

    return ts.tv_sec + ts.tv_nsec * 1e-9;
}
```

macOS version:
```C
void criterion_inittime(void)
{
    if (timebase_recip == 0) {
	mach_timebase_info(&timebase_info);
	timebase_recip = (timebase_info.denom / timebase_info.numer) / 1e9;
    }
}

double criterion_gettime(void)
{
    return mach_absolute_time() * timebase_recip;
}
```

Running `initializeTime` before any benchmarks fixes the issue in my minimal example. Should I make a PR that just adds an `initializeTime` line to `runBenchmark`?

**Comment by RyanGlScott** [comment]
Oh dear. That certainly explains things, alright.

While your suggested fix would certainly work, you've made me realize that there are other `IO`-returning functions in `criterion` that likely suffer from this issue (e.g., `measure`). This leaves me wondering how we should deal with this issue more generally. I can think of three possible paths forward:

1. Perform an audit of `criterion` to find `IO`-returning functions that call `getTime` without calling `initializeTime` beforehand, and fix them.
2. Document this infelicity. Currently, the Haddocks for `getTime` state that it is the users' responsibility to call `initializeTime` beforehand. Given that `runBenchmark`, `measure`, etc. are in some sense more low-level than `defaultMain`, perhaps we should also make it the users' responsibility to call `initializeTime` before invoking any of these other functions as well.
3. Having `getTime` always call `initializeTime` beforehand. This sounds strange, but it would be doable, since `initializeTime` is idempotent, so we'd only ever actually initialize the timer once in a `criterion` programs' execution.


Each option has their pros and cons:

1. This option wouldn't require any downstream clients to change their code. However, this approach would likely be the most error-prone, however, since refactoring code in the future could likely open up new ways to invoke `getTime` without `initializeTime`.
2. If we can't be confident that we'd cover every possible gap, then this option is attractive, since it shifts the burden of calling `initializeTime` beforehand onto users who wish to use lower-level functionality. On the other hand, it would require users to update their code (and only to fix a bug on non-Linux operating systems, at that).
3. This would be a very straightforward fix, and it wouldn't require users to change any code. On the other hand, it's a bit risky in that there is some runtime cost to actually initializing a timer, and if users' aren't careful, they could incur it while measuring something else. Perhaps this runtime cost is negligible enough that it can be ignored, though? (_à la_ [`withSocketsDo`](https://github.com/haskell/network/pull/158) from `network`).

What do others think?

**Comment by doomrobo** [comment]
I think option 3 wouldn't be great, since it would add overhead to every call to `getTime` in a hot benchmarking loop, which might skew results.

Looking at the codebase, it appears that `getTime` is only called by `measure` and `runBenchmark` in `Criterion.Measurement` (it's also called in `examples/Overhead.hs`). I think it'd be most ergonomic to add an `initializeTime` to both of those, since it wouldn't add much overhead, and *also* add a notice to the `getTime` function.

We can also update `examples/Overhead.hs` to reflect this. That'd actually make for a nice example that newbies can stumble upon.

I can write this up and submit a PR if this sounds good to you.

**Comment by RyanGlScott** [comment]
I'd be fine with your suggestion, so a pull request implementing that would be much appreciated.

**Comment by RyanGlScott** [comment]
Fixed in #196.
